src/controllers/nodes/nodes_controller.go

```diff
@@ -10,235 +10,198 @@ import (
-				return err
-			}
+	// Node is found in the cluster and in cache, send mark for termination, not found node is handled in err check
+	if dynakube != nil {
+		cachedNodeData := CachedNodeInfo{
+			cachedNode: cachedNodeInfo,
+			nodeCache:  nodeCache,
+			nodeName:   nodeName,
 		}
-	}
-
-	// Add or update all nodes seen on OneAgent instances to the c.
-	for _, dk := range dks {
-		if dk.Status.OneAgent.Instances != nil {
-			for node, info := range dk.Status.OneAgent.Instances {
-				if _, ok := nodes[node]; !ok {
-					continue
-				}
-
-				info := CacheEntry{
-					Instance:  dk.Name,
-					IPAddress: info.IPAddress,
-					LastSeen:  time.Now().UTC(),
-				}
 
-				if cached, err := c.Get(node); err == nil {
-					info.LastMarkedForTermination = cached.LastMarkedForTermination
-				}
-
-				if err := c.Set(node, info); err != nil {
-					return err
-				}
-			}
+		if err := r.markForTermination(dynakube, cachedNodeData); err != nil {
+			log.Error(err, "error while sending mark for termination for node:", "node", nodeName)
+			return err
 		}
 	}
 
-	// Notify and remove all nodes on the c that aren't in the cluster.
-	for _, node := range c.Keys() {
-		if _, ok := nodes[node]; ok {
-			continue
-		}
-
-		if err := controller.removeNode(c, node, func(name string) (*dynatracev1beta1.DynaKube, error) {
-			if dk, ok := dks[name]; ok {
-				return dk, nil
-			}
-
-			return nil, errors.NewNotFound(schema.GroupResource{
-				Group:    dkList.GroupVersionKind().Group,
-				Resource: dkList.GroupVersionKind().Kind,
-			}, name)
-		}); err != nil {
-			log.Error(err, "failed to remove node", "node", node)
-		}
+	nodeCache.Delete(nodeName)
+	if err := r.updateCache(nodeCache, context.TODO()); err != nil {
+		log.Error(err, "error while updating node cache after deletion:", "node", nodeName)
+		return err
 	}
-
-	return controller.updateCache(c)
+	return nil
 }
 
-func (controller *NodesController) getCache() (*Cache, error) {
+func (r *ReconcileNode) getCache() (*Cache, error) {
 	var cm corev1.ConfigMap
 
-	err := controller.client.Get(context.TODO(), client.ObjectKey{Name: cacheName, Namespace: controller.namespace}, &cm)
+	err := r.client.Get(context.TODO(), client.ObjectKey{Name: cacheName, Namespace: r.podNamespace}, &cm)
 	if err == nil {
 		return &Cache{Obj: &cm}, nil
 	}
 
-	if errors.IsNotFound(err) {
-		log.Info("no cache found, creating")
-
+	if k8serrors.IsNotFound(err) {
 		cm := &corev1.ConfigMap{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      cacheName,
-				Namespace: controller.namespace,
+				Namespace: r.podNamespace,
 			},
 			Data: map[string]string{},
 		}
 
-		if !controller.local { // If running locally, don't set the controller.
-			deploy, err := kubeobjects.GetDeployment(controller.client, controller.podName, controller.namespace)
+		if !r.runLocal { // If running locally, don't set the controller.
+			deploy, err := kubeobjects.GetDeployment(r.client, os.Getenv("POD_NAME"), r.podNamespace)
```

Comment:
`os.Getenv()` here makes unit testing weird.
better to pass it in via the `ReconcileNodes` struct

Reply:
Unit Tests will never reach that point, as runLocal is true, when unit testing, thats why os.Getenv() can be used here